server/src-lib/Hasura/GraphQL/Resolve/Mutation.hs

```diff
@@ -107,35 +112,60 @@ parseConstraint obj = do
 
 parseOnConflict
   :: (MonadError QErr m)
-  => [PGCol] -> AnnGValue -> m RI.ConflictClauseP1
-parseOnConflict cols val =
+  => AnnGValue -> m ConflictCtx
+parseOnConflict val =
   flip withObject val $ \_ obj -> do
     action <- parseAction obj
     constraintM <- parseConstraint obj
-    mkConflictClause cols action constraintM
+    return (action, constraintM)
 
 convertInsert
-  :: (QualifiedTable, QualifiedTable) -- table, view
+  :: RoleName
+  -> (QualifiedTable, QualifiedTable) -- table, view
   -> [PGCol] -- all the columns in this table
   -> Field -- the mutation field
   -> Convert RespTx
-convertInsert (tn, vn) tableCols fld = do
+convertInsert role (tn, vn) tableCols fld = do
   rows    <- withArg arguments "objects" asRowExps
-  onConflictM <- withPathK "on_conflict" $
-                 withArgM arguments "on_conflict" $
-                 parseOnConflict tableCols
+  conflictCtxM <- withPathK "on_conflict" $
+                 withArgM arguments "on_conflict" parseOnConflict
   mutFlds <- convertMutResp (_fType fld) $ _fSelSet fld
   args <- get
-  let p1 = RI.InsertQueryP1 tn vn tableCols rows onConflictM mutFlds
-  return $ RI.insertP2 (p1, args)
+  return $
+    bool (nonAdminInsert args rows conflictCtxM mutFlds)
+         (adminInsert args rows conflictCtxM mutFlds)
+         $ isAdmin role
   where
     arguments = _fArguments fld
     asRowExps = withArray (const $ mapM rowExpWithDefaults)
     rowExpWithDefaults val = do
       givenCols <- convertRowObj val
       return $ Map.elems $ Map.union (Map.fromList givenCols) defVals
+
     defVals = Map.fromList $ zip tableCols (repeat $ S.SEUnsafe "DEFAULT")
 
+    adminInsert args rows conflictCtxM mutFlds = do
+      onConflictM <- mapM (mkConflictClause tableCols) conflictCtxM
+      let p1 = RI.InsertQueryP1 tn vn tableCols rows onConflictM mutFlds
+      RI.insertP2 (p1, args)
+
+    nonAdminInsert args rows conflictCtxM mutFlds = do
+      mapM_ (mkConflictClause tableCols) conflictCtxM
+      setConflictCtxTx conflictCtxM
+      let p1 = RI.InsertQueryP1 tn vn tableCols rows Nothing mutFlds
+      RI.insertP2 (p1, args)
+
+    setConflictCtxTx conflictCtxM = do
+      let t = maybe "{}" conflictCtxToJSON conflictCtxM
+          setVal = toSQL $ S.SELit t
+          setVar = BB.string7 "SET LOCAL hasura.conflict_clause = "
+          q = Q.fromBuilder $ setVar <> setVal
+      Q.unitQE defaultTxErrorHandler q () False
+
+    conflictCtxToJSON (act, constrM) =
+      let ctx = InsertTxConflictCtx act constrM
+      in CS.cs $ A.encode $ A.object ["conflict_clause" A..= ctx]
```

Comment:
https://hackage.haskell.org/package/aeson-1.0.2.1/docs/Data-Aeson-Text.html#v:encodeToLazyText

and convert it to strict text instead of using `cs`.